### ISSUE TITLE
Bug: [no-unnecessary-parameter-property-assignment] false positive on computed access with a variable key

### ISSUE THREAD

**Issue by zaewc** [typescript-eslint/typescript-eslint#12566]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=MYGwhgzhAECC0G8BQ1rAPYDsIBcBOArsDungBQAOBARiAJbDQBm66AXNLnnZgOYCUiFKmg4AFnQgBtFugC60ALzNWAbmEBfJBqA&eslintrc=N4KABARATgrgNgUwM4QFxlGSABALgTwAdkBjKAS0NwFpk5yA7XAegYHtqYGGETkkAhlHzVCQgQFsEuBFFFQ2xKAWoCkScgHMGUpmkiyFUCODABfEGaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class A {
  constructor(public foo: string) {
    this[foo] = foo;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-parameter-property-assignment": "error",
  },
};
```

### tsconfig

```jsonc
na
```

### Expected Result

No error

`this[foo]` uses the runtime value of foo as the property key, so it is not statically equivalent to `this.foo`

### Actual Result

`this[foo] = foo` is reported:
> This assignment is unnecessary since it is already assigned by a parameter property. 3:5 - 3:20

### Additional Info

I'm happy to open a PR
